File: src/Application Specific/Classes/Entity Collections/Contact/ContactSQL.vb
Imports System.Data.SqlClient

Namespace Entity

    Namespace Contacts

        Public Class ContactSQL
            Private _database As Sys.Database

            Public Sub New(ByVal database As Sys.Database)
                _database = database
            End Sub

#Region "Functions"

            Public Sub Delete(ByVal ContactID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@ContactID", ContactID, True)
                _database.ExecuteSP_NO_Return("Contact_Delete")
            End Sub

            Public Function [Insert](ByVal oContact As Contact) As Contact
                _database.ClearParameter()
                AddContactParametersToCommand(oContact)

                oContact.SetContactID = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Contact_Insert"))
                oContact.Exists = True
                Return oContact
            End Function

            Public Function [Contact_Search](ByVal criteria As String) As DataView
                _database.ClearParameter()
                _database.AddParameter("@Criteria", criteria, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("Contact_Search")
                dt.TableName = Entity.Sys.Enums.TableNames.tblContact.ToString
                Return New DataView(dt)
            End Function

            Public Sub [Update](ByVal oContact As Contact)
                _database.ClearParameter()
                _database.AddParameter("@ContactID", oContact.ContactID, True)
                AddContactParametersToCommand(oContact)
                _database.ExecuteSP_NO_Return("Contact_Update")
            End Sub

            Public Function Contact_GetForSite(ByVal SiteID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@SiteID", SiteID, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Contact_GetForSite")
                dt.TableName = Entity.Sys.Enums.TableNames.tblContact.ToString
                Return New DataView(dt)
            End Function

            Public Function Contact_GetAll() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Contact_GetAll")
                dt.TableName = Entity.Sys.Enums.TableNames.tblContact.ToString
                Return New DataView(dt)
            End Function

            Public Function Contact_Get(ByVal ContactID As Integer) As Contact
                _database.ClearParameter()
                _database.AddParameter("@ContactID", ContactID)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Contact_Get")

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oContact As New Contact
                        With oContact
                            .IgnoreExceptionsOnSetMethods = True
                            .SetContactID = dt.Rows(0).Item("ContactID")
                            .SetSalutation = dt.Rows(0).Item("Salutation")
                            .SetFirstName = dt.Rows(0).Item("FirstName")
                            .SetSurname = dt.Rows(0).Item("Surname")
                            .SetEmailAddress = dt.Rows(0).Item("EmailAddress")
                            .SetTelephoneNum = dt.Rows(0).Item("TelephoneNum")
                            .SetFaxNum = dt.Rows(0).Item("FaxNum")
                            .SetPropertyID = dt.Rows(0).Item("SiteID")
                            .SetDeleted = dt.Rows(0).Item("Deleted")
                            .SetMobileNo = dt.Rows(0).Item("MobileNo")
                            .SetPortalUserName = dt.Rows(0).Item("PortalUserName")
                            .SetPortalPassword = dt.Rows(0).Item("PortalPassword")
                            .SetPortalGSRPrint = dt.Rows(0).Item("PortalGSRPrint")

                            .SetEID = dt.Rows(0).Item("EID")
                        End With
                        oContact.Exists = True
                        Return oContact
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function Check_Unique_PortalUsername(ByVal username As String, ByVal ID As Integer) As Boolean
                Dim NumberOfUsers As Integer = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteScalar("SELECT COUNT(Contactid) as 'NumberOfUsers' " &
                "FROM tblContact WHERE portalusername = '" & username & "' AND contactid <> " & ID & " AND PortalUserName IS NOT NULL", False))

                If NumberOfUsers = 0 Then
                    Return True
                Else
                    Return False
                End If
            End Function

            Private Sub AddContactParametersToCommand(ByRef oContact As Contact)
                With _database
                    .AddParameter("@FirstName", oContact.FirstName, True)
                    .AddParameter("@Salutation", oContact.Salutation, True)
                    .AddParameter("@Surname", oContact.Surname, True)
                    .AddParameter("@TelephoneNum", oContact.TelephoneNum, True)
                    .AddParameter("@EmailAddress", oContact.EmailAddress, True)
                    .AddParameter("@FaxNum", oContact.FaxNum, True)
                    .AddParameter("@SiteID", oContact.PropertyID, True)
                    .AddParameter("@MobileNo", oContact.MobileNo, True)
                    If Not oContact.PortalUserName.Trim.Length = 0 Then
                        .AddParameter("@PortalUserName", oContact.PortalUserName, True)
                    End If
                    If Not oContact.PortalPassword.Trim.Length = 0 Then
                        .AddParameter("@PortalPassword", oContact.PortalPassword, True)
                    End If
                    .AddParameter("@PortalGSRPrint", oContact.PortalGSRPrint, True)
                    If Not oContact.EID = 0 Then
                        .AddParameter("@EID", oContact.EID, True)
                    End If
                End With
            End Sub

#End Region

#Region "Contacts"

            Public Function [Contacts_GetAll_ForLink](ByVal linkId As Integer, ByVal linkRef As Integer,
                                                      Optional ByVal deleted As Boolean = False) As DataView
                _database.ClearParameter()
                _database.AddParameter("@LinkRef", linkRef, True)
                _database.AddParameter("@LinkID", linkId, True)
                _database.AddParameter("@Deleted", deleted, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Contacts_GetAll_ForLink")
                dt.TableName = Entity.Sys.Enums.TableNames.tblContact.ToString
                Return New DataView(dt)
            End Function

            Public Function [Contacts_Get](ByVal contactId As Integer) As Contact
                _database.ClearParameter()
                _database.AddParameter("@ContactID", contactId)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Contacts_Get")
                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oContact As New Contact
                        With oContact
                            .IgnoreExceptionsOnSetMethods = True
                            If dt.Columns.Contains("ContactID") Then .SetContactID = dt.Rows(0).Item("ContactID")
                            If dt.Columns.Contains("Title") Then .SetSalutation = dt.Rows(0).Item("Title")
                            If dt.Columns.Contains("FirstName") Then .SetFirstName = dt.Rows(0).Item("FirstName")
                            If dt.Columns.Contains("LastName") Then .SetSurname = dt.Rows(0).Item("LastName")
                            If dt.Columns.Contains("Address1") Then .SetAddress1 = dt.Rows(0).Item("Address1")
                            If dt.Columns.Contains("Address2") Then .SetAddress2 = dt.Rows(0).Item("Address2")
                            If dt.Columns.Contains("Address3") Then .SetAddress3 = dt.Rows(0).Item("Address3")
                            If dt.Columns.Contains("Postcode") Then .SetPostcode = dt.Rows(0).Item("Postcode")
                            If dt.Columns.Contains("EmailAddress") Then .SetEmailAddress = dt.Rows(0).Item("EmailAddress")
                            If dt.Columns.Contains("TelephoneNo") Then .SetTelephoneNum = dt.Rows(0).Item("TelephoneNo")
                            If dt.Columns.Contains("MobileNo") Then .SetMobileNo = dt.Rows(0).Item("MobileNo")
                            If dt.Columns.Contains("LinkID") Then .SetLinkID = dt.Rows(0).Item("LinkID")
                            If dt.Columns.Contains("LinkRef") Then .SetLinkRef = dt.Rows(0).Item("LinkRef")
                            If dt.Columns.Contains("NoMarketing") Then .SetNoMarketing = dt.Rows(0).Item("NoMarketing")
                            If dt.Columns.Contains("OnContract") Then .SetOnContract = dt.Rows(0).Item("OnContract")
                            If dt.Columns.Contains("RelationshipID") Then .SetRelationshipID = Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("RelationshipID"))
                            If dt.Columns.Contains("POC") Then .SetPointOfContact = Entity.Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("POC"))
                            If dt.Columns.Contains("Relationship") Then .SetRelationship = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("Relationship"))
                            If dt.Columns.Contains("Deleted") Then .SetDeleted = dt.Rows(0).Item("Deleted")
                        End With
                        oContact.Exists = True
                        Return oContact
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function [Contacts_Update](ByVal oContact As Contact) As Integer
                _database.ClearParameter()
                _database.AddParameter("@ContactID", oContact.ContactID, True)
                _database.AddParameter("@Title", oContact.Salutation, True)
                _database.AddParameter("@FirstName", oContact.FirstName, True)
                _database.AddParameter("@LastName", oContact.Surname, True)
                _database.AddParameter("@TelephoneNo", oContact.TelephoneNum, True)
                _database.AddParameter("@MobileNo", oContact.MobileNo, True)
                _database.AddParameter("@Address1", oContact.Address1, True)
                _database.AddParameter("@Address2", oContact.Address2, True)
                _database.AddParameter("@Address3", oContact.Address3, True)
                _database.AddParameter("@Postcode", oContact.Postcode, True)
                _database.AddParameter("@EmailAddress", oContact.EmailAddress, True)
                _database.AddParameter("@NoMarketing", oContact.NoMarketing, True)
                _database.AddParameter("@LinkID", oContact.LinkID, True)
                _database.AddParameter("@LinkRef", oContact.LinkRef, True)
                _database.AddParameter("@OnContract", oContact.OnContract, True)
                _database.AddParameter("@RelationshipID", oContact.RelationshipID, True)
                _database.AddParameter("@POC", oContact.PointOfContact, True)
                Return Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Contacts_Update"))
            End Function

            Public Function [Contacts_Delete](ByVal contactId As Integer) As Boolean
                _database.ClearParameter()
                _database.AddParameter("@ContactID", contactId, True)
                Return CBool(_database.ExecuteSP_ReturnRowsAffected("Contacts_Delete") = 1)

            End Function

#End Region

        End Class

    End Namespace

End Namespace

File: apollo/apollo.fsm/src/Application Specific/Classes/Entity Collections/Contact/ContactSQL.cs
using System.Data;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace Contacts
    {
        public class ContactSQL
        {
            private Sys.Database _database;

            public ContactSQL(Sys.Database database)
            {
                _database = database;
            }

            public void Delete(int ContactID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContactID", ContactID, true);
                _database.ExecuteSP_NO_Return("Contact_Delete");
            }

            public Contact Insert(Contact oContact)
            {
                _database.ClearParameter();
                AddContactParametersToCommand(ref oContact);
                oContact.SetContactID = Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Contact_Insert"));
                oContact.Exists = true;
                return oContact;
            }

            public void Update(Contact oContact)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContactID", oContact.ContactID, true);
                AddContactParametersToCommand(ref oContact);
                _database.ExecuteSP_NO_Return("Contact_Update");
            }

            public DataView Contact_GetForSite(int SiteID)
            {
                _database.ClearParameter();
                _database.AddParameter("@SiteID", SiteID, true);
                var dt = _database.ExecuteSP_DataTable("Contact_GetForSite");
                dt.TableName = Sys.Enums.TableNames.tblContact.ToString();
                return new DataView(dt);
            }

            public DataView Contact_GetAll()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("Contact_GetAll");
                dt.TableName = Sys.Enums.TableNames.tblContact.ToString();
                return new DataView(dt);
            }

            public Contact Contact_Get(int ContactID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContactID", ContactID);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("Contact_Get");
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oContact = new Contact();
                        oContact.IgnoreExceptionsOnSetMethods = true;
                        oContact.SetContactID = dt.Rows[0]["ContactID"];
                        oContact.SetSalutation = dt.Rows[0]["Salutation"];
                        oContact.SetFirstName = dt.Rows[0]["FirstName"];
                        oContact.SetSurname = dt.Rows[0]["Surname"];
                        oContact.SetEmailAddress = dt.Rows[0]["EmailAddress"];
                        oContact.SetTelephoneNum = dt.Rows[0]["TelephoneNum"];
                        oContact.SetFaxNum = dt.Rows[0]["FaxNum"];
                        oContact.SetPropertyID = dt.Rows[0]["SiteID"];
                        oContact.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oContact.SetMobileNo = dt.Rows[0]["MobileNo"];
                        oContact.SetPortalUserName = dt.Rows[0]["PortalUserName"];
                        oContact.SetPortalPassword = dt.Rows[0]["PortalPassword"];
                        oContact.SetPortalGSRPrint = dt.Rows[0]["PortalGSRPrint"];
                        oContact.SetEID = dt.Rows[0]["EID"];
                        oContact.Exists = true;
                        return oContact;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            private void AddContactParametersToCommand(ref Contact oContact)
            {
                {
                    var withBlock = _database;
                    withBlock.AddParameter("@FirstName", oContact.FirstName, true);
                    withBlock.AddParameter("@Salutation", oContact.Salutation, true);
                    withBlock.AddParameter("@Surname", oContact.Surname, true);
                    withBlock.AddParameter("@TelephoneNum", oContact.TelephoneNum, true);
                    withBlock.AddParameter("@EmailAddress", oContact.EmailAddress, true);
                    withBlock.AddParameter("@FaxNum", oContact.FaxNum, true);
                    withBlock.AddParameter("@SiteID", oContact.PropertyID, true);
                    withBlock.AddParameter("@MobileNo", oContact.MobileNo, true);
                    if (!(oContact.PortalUserName.Trim().Length == 0))
                    {
                        withBlock.AddParameter("@PortalUserName", oContact.PortalUserName, true);
                    }

                    if (!(oContact.PortalPassword.Trim().Length == 0))
                    {
                        withBlock.AddParameter("@PortalPassword", oContact.PortalPassword, true);
                    }

                    withBlock.AddParameter("@PortalGSRPrint", oContact.PortalGSRPrint, true);
                    if (!(oContact.EID == 0))
                    {
                        withBlock.AddParameter("@EID", oContact.EID, true);
                    }
                }
            }

            public DataView Contacts_GetAll_ForLink(int linkId, int linkRef, bool deleted = false)
            {
                _database.ClearParameter();
                _database.AddParameter("@LinkRef", linkRef, true);
                _database.AddParameter("@LinkID", linkId, true);
                _database.AddParameter("@Deleted", deleted, true);
                var dt = _database.ExecuteSP_DataTable("Contacts_GetAll_ForLink");
                dt.TableName = Sys.Enums.TableNames.tblContact.ToString();
                return new DataView(dt);
            }

            public Contact Contacts_Get(int contactId)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContactID", contactId);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("Contacts_Get");
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oContact = new Contact();
                        oContact.IgnoreExceptionsOnSetMethods = true;
                        if (dt.Columns.Contains("ContactID"))
                            oContact.SetContactID = dt.Rows[0]["ContactID"];
                        if (dt.Columns.Contains("Title"))
                            oContact.SetSalutation = dt.Rows[0]["Title"];
                        if (dt.Columns.Contains("FirstName"))
                            oContact.SetFirstName = dt.Rows[0]["FirstName"];
                        if (dt.Columns.Contains("LastName"))
                            oContact.SetSurname = dt.Rows[0]["LastName"];
                        if (dt.Columns.Contains("Address1"))
                            oContact.SetAddress1 = dt.Rows[0]["Address1"];
                        if (dt.Columns.Contains("Address2"))
                            oContact.SetAddress2 = dt.Rows[0]["Address2"];
                        if (dt.Columns.Contains("Address3"))
                            oContact.SetAddress3 = dt.Rows[0]["Address3"];
                        if (dt.Columns.Contains("Postcode"))
                            oContact.SetPostcode = dt.Rows[0]["Postcode"];
                        if (dt.Columns.Contains("EmailAddress"))
                            oContact.SetEmailAddress = dt.Rows[0]["EmailAddress"];
                        if (dt.Columns.Contains("TelephoneNo"))
                            oContact.SetTelephoneNum = dt.Rows[0]["TelephoneNo"];
                        if (dt.Columns.Contains("MobileNo"))
                            oContact.SetMobileNo = dt.Rows[0]["MobileNo"];
                        if (dt.Columns.Contains("LinkID"))
                            oContact.SetLinkID = dt.Rows[0]["LinkID"];
                        if (dt.Columns.Contains("LinkRef"))
                            oContact.SetLinkRef = dt.Rows[0]["LinkRef"];
                        if (dt.Columns.Contains("NoMarketing"))
                            oContact.SetNoMarketing = Conversions.ToBoolean(dt.Rows[0]["NoMarketing"]);
                        if (dt.Columns.Contains("OnContract"))
                            oContact.SetOnContract = Conversions.ToBoolean(dt.Rows[0]["OnContract"]);
                        if (dt.Columns.Contains("RelationshipID"))
                            oContact.SetRelationshipID = Sys.Helper.MakeIntegerValid(dt.Rows[0]["RelationshipID"]);
                        if (dt.Columns.Contains("POC"))
                            oContact.SetPointOfContact = Sys.Helper.MakeBooleanValid(dt.Rows[0]["POC"]);
                        if (dt.Columns.Contains("Relationship"))
                            oContact.SetRelationship = Conversions.ToInteger(Sys.Helper.MakeStringValid(dt.Rows[0]["Relationship"]));
                        if (dt.Columns.Contains("Deleted"))
                            oContact.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oContact.Exists = true;
                        return oContact;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public int Contacts_Update(Contact oContact)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContactID", oContact.ContactID, true);
                _database.AddParameter("@Title", oContact.Salutation, true);
                _database.AddParameter("@FirstName", oContact.FirstName, true);
                _database.AddParameter("@LastName", oContact.Surname, true);
                _database.AddParameter("@TelephoneNo", oContact.TelephoneNum, true);
                _database.AddParameter("@MobileNo", oContact.MobileNo, true);
                _database.AddParameter("@Address1", oContact.Address1, true);
                _database.AddParameter("@Address2", oContact.Address2, true);
                _database.AddParameter("@Address3", oContact.Address3, true);
                _database.AddParameter("@Postcode", oContact.Postcode, true);
                _database.AddParameter("@EmailAddress", oContact.EmailAddress, true);
                _database.AddParameter("@NoMarketing", oContact.NoMarketing, true);
                _database.AddParameter("@LinkID", oContact.LinkID, true);
                _database.AddParameter("@LinkRef", oContact.LinkRef, true);
                _database.AddParameter("@OnContract", oContact.OnContract, true);
                _database.AddParameter("@RelationshipID", oContact.RelationshipID, true);
                _database.AddParameter("@POC", oContact.PointOfContact, true);
                return Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Contacts_Update"));
            }

            public bool Contacts_Delete(int contactId)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContactID", contactId, true);
                return Conversions.ToBoolean(_database.ExecuteSP_ReturnRowsAffected("Contacts_Delete") == 1);
            }
        }
    }
}